src/utils/makeReactNativeConfig.js

```diff
@@ -224,13 +230,26 @@ function makeReactNativeConfig(
   const defaultWebpackConfig = getDefaultConfig(env);
   const polyfillPath = require.resolve('./polyfillEnvironment.js');
 
-  const userConfig =
-    typeof userWebpackConfig === 'function'
-      ? userWebpackConfig(env, defaultWebpackConfig)
-      : userWebpackConfig;
+  /**
+  * Load config from file. Support both - "module.exports" and "export default {}"
+  */
+  const { webpack: webpackConfigFactory /* ...haulConfig */ } =
+    userWebpackConfig.default || userWebpackConfig;
+
+  if (
+    typeof webpackConfigFactory !== 'function' &&
+    typeof webpackConfigFactory !== 'object'
+  ) {
+    throw new Error('You have provided a wrong Webpack configuration.');
+  }
+
+  const webpackConfig =
+    typeof webpackConfigFactory === 'function'
+      ? webpackConfigFactory(env)
+      : webpackConfigFactory;
 
-  const config = Object.assign({}, defaultWebpackConfig, userConfig, {
-    entry: injectPolyfillIntoEntry(userConfig.entry, polyfillPath),
+  const config = merge(defaultWebpackConfig, webpackConfig, {
```

Comment:
I don't think we should be deep merging webpack configs. This could produce weird results. Also the config received here should already a full webpack config and we shouldn't do any changes to that (see below).

Reply:
That doesn't make any damn sense.
> The createWebpackConfig utility can take some common options has return a webpack configuration function which takes the env argument. 

What common options? What are those? Either you need to be more specific in the proposal or this should be left as it is.